src/style/style_layer/fill_extrusion_style_layer.js

```diff
@@ -220,4 +253,96 @@ function projectExtrusion(geometry: Array<Array<Point>>, zBase: number, zTop: nu
     return [projectedBase, projectedTop];
 }
 
+/**
+ * Projects a fill extrusion vertices to screen while accounting for terrain.
+ * This and its dependent functions are ported directly from `fill_extrusion.vertex.glsl`
+ * with a few co-ordinate space differences.
+ *
+ * - Matrix `m` projects to screen-pixel space instead of to gl-coordinates (NDC)
+ * - Texture querying is performed in texture pixel coordinates instead of  normalized uv coordinates.
+ * - Height offset calculation for fill-extrusion-base is offset with -1 instead of -5 to prevent underground picking.
+ */
+function projectExtrusion3D(geometry: Array<Array<Point>>, zBase: number, zTop: number, translation: Point, m: Float32Array, demSampler: DEMSampler, centroid: vec2, exaggeration: number, lat: number) {
+    const projectedBase = [];
+    const projectedTop = [];
+
+    for (const r of geometry) {
+        const ringBase = [];
+        const ringTop = [];
+        for (const p of r) {
+            const x = p.x + translation.x;
+            const y = p.y + translation.y;
+            const heightOffset = getTerrainHeightOffset(x, y, zBase, zTop, demSampler, centroid, exaggeration, lat);
+
+            const base = toPoint(vec3.transformMat4([], [x, y, heightOffset.base], m));
+            const top = toPoint(vec3.transformMat4([], [x, y, heightOffset.top], m));
```

Comment:
We can get rid of two array allocations per point here by doing in-place matrix transforms (`vec3.transformMat4(base, base, m)`).